module/scmi_reset/src/mod_scmi_reset.c

```diff
@@ -0,0 +1,447 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2015-2019, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     SCMI Reset Domain management protocol support.
+ */
+
+#include <string.h>
+#include <fwk_assert.h>
+#include <fwk_element.h>
+#include <fwk_errno.h>
+#include <fwk_id.h>
+#include <fwk_macros.h>
+#include <fwk_module.h>
+#include <fwk_module_idx.h>
+#include <internal/scmi.h>
+#include <internal/scmi_reset.h>
+#include <mod_log.h>
+#include <mod_reset.h>
+#include <mod_scmi.h>
+#include <mod_scmi_reset.h>
+
+struct scmi_reset_ctx {
+    /*! SCMI Reset Module Configuration */
+    const struct mod_scmi_reset_config *config;
+    /* Table of agent descriptors,for per-agent views of reset domains */
+    const struct mod_scmi_reset_agent *agent_table;
+    /* Binded module APIs */
+    const struct mod_log_api *log_api;
+    const struct mod_scmi_from_protocol_api *scmi_api;
+    const struct mod_reset_drv_api *reset_api;
+};
+
+static int protocol_version_handler(fwk_id_t service_id,
+                                    const uint32_t *payload);
+
+static int protocol_attributes_handler(fwk_id_t service_id,
+                                       const uint32_t *payload);
+
+static int protocol_message_attributes_handler(fwk_id_t service_id,
+                                               const uint32_t *payload);
+
+static int reset_attributes_handler(fwk_id_t service_id,
+                                    const uint32_t *payload);
+
+static int reset_request_handler(fwk_id_t service_id,
+                                 const uint32_t *payload);
+
+static int reset_notify_handler(fwk_id_t service_id,
+                                const uint32_t *payload);
+
+/*
+ * Internal variables
+ */
+
+static struct scmi_reset_ctx scmi_reset_ctx;
+
+static int (*msg_handler_table[])(fwk_id_t, const uint32_t *) = {
+    [SCMI_PROTOCOL_VERSION] = protocol_version_handler,
+    [SCMI_PROTOCOL_ATTRIBUTES] = protocol_attributes_handler,
+    [SCMI_PROTOCOL_MESSAGE_ATTRIBUTES] = protocol_message_attributes_handler,
+    [SCMI_RESET_DOMAIN_ATTRIBUTES] = reset_attributes_handler,
+    [SCMI_RESET_REQUEST] = reset_request_handler,
+    [SCMI_RESET_NOTIFY] = reset_notify_handler,
+};
+
+static unsigned int payload_size_table[] = {
+    [SCMI_PROTOCOL_VERSION] = 0,
+    [SCMI_PROTOCOL_ATTRIBUTES] = 0,
+    [SCMI_PROTOCOL_MESSAGE_ATTRIBUTES] =
+                       sizeof(struct scmi_protocol_message_attributes_a2p),
+    [SCMI_RESET_DOMAIN_ATTRIBUTES] = sizeof(struct scmi_reset_attributes_a2p),
+    [SCMI_RESET_REQUEST] = sizeof(struct scmi_reset_request_a2p),
+    [SCMI_RESET_NOTIFY] = sizeof(struct scmi_reset_notify_a2p),
+};
+
+static unsigned int count_domains(void)
+{
+    unsigned int i = 0;
+    unsigned int count = 0;
+
+    for (i = 0; i < scmi_reset_ctx.config->agent_count; i++)
+	count += scmi_reset_ctx.config->agent_table[i].device_count;
```

Comment:
This looks flawed to me. What if we have 2 agents and both of them have access to all the reset domains or there is a reset domain shared among 2 agents. We would count more in this case. This needs to return the actual number of reset domains.

Reply:
ok, i need to check this.

Reply:
addressed in latest commits post.